sdk/schemaregistry/schema-registry-avro-serializer/src/schemaRegistryAvroSerializer.ts

```diff
@@ -0,0 +1,230 @@
+  /**
+   * Serializes a value into a buffer.
+   *
+   * @param value The value to serialize.
+   * @param schema The Avro schema to use.
+   * @returns A new buffer with the serialized value
+   */
+  async serialize(value: any, schema: string): Promise<Buffer> {
+    const entry = await this.getSchemaByContent(schema);
+    const payload = entry.type.toBuffer(value);
+    const buffer = Buffer.alloc(PAYLOAD_OFFSET + payload.length);
+
+    buffer.writeUInt32BE(FORMAT_INDICATOR, 0);
+    buffer.write(entry.id, SCHEMA_ID_OFFSET, SCHEMA_ID_LENGTH, "utf-8");
+    payload.copy(buffer, PAYLOAD_OFFSET);
+    return buffer;
+  }
+
+  // REVIEW: signature. See serialize and s/serialize into/deserialize from/.
+  /**
+   * Deserializes a value from a buffer.
+   *
+   * @param buffer The buffer with the serialized value.
+   * @return The deserialized value.
+   */
+  async deserialize(buffer: Buffer): Promise<any> {
+    if (buffer.length < PAYLOAD_OFFSET) {
+      throw new Error("Buffer is too small to have the correct format.");
+    }
+
+    const format = buffer.readUInt32BE(0);
+    if (format !== FORMAT_INDICATOR) {
+      throw new Error(`Buffer has unknown format indicator: 0x${format.toString(16)}`);
+    }
+
+    const schemaIdBuffer = buffer.slice(SCHEMA_ID_OFFSET, PAYLOAD_OFFSET);
+    const schemaId = schemaIdBuffer.toString("utf-8");
+    const schema = await this.getSchemaById(schemaId);
+    const payloadBuffer = buffer.slice(PAYLOAD_OFFSET);
+
+    return schema.type.fromBuffer(payloadBuffer);
+  }
+
+  //
+  // Avoid hitting the schema registry on every call by caching schemas.
+  //
+  // Currently the cache can only be discarded by throwing away the serializer
+  // and creating a new one, but this will be revisited after the initial
+  // preview:
+  //
+  // https://github.com/Azure/azure-sdk-for-js/issues/10438
+  //
+  private readonly cacheByContent = new Map<string, CacheEntry>();
+  private readonly cacheById = new Map<string, CacheEntry>();
+
+  private async getSchemaById(schemaId: string): Promise<CacheEntry> {
+    const cached = this.cacheById.get(schemaId);
+    if (cached) {
+      return cached;
+    }
+
+    const schemaResponse = await this.registry.getSchemaById(schemaId);
+    if (schemaResponse.serializationType != "avro") {
+      throw new Error(
+        `Schema with ID '${schemaResponse.id}' has has serialization type '${schemaResponse.serializationType}', not 'avro'.`
+      );
+    }
+
+    const avroType = avro.Type.forSchema(JSON.parse(schemaResponse.content));
+    return this.cache(schemaId, schemaResponse.content, avroType);
+  }
+
+  private async getSchemaByContent(schema: string): Promise<CacheEntry> {
+    let cached = this.cacheByContent.get(schema);
+    if (cached) {
+      return cached;
+    }
+
+    // REVIEW: normalize string here? It was my understanding that the schema
+    //         registry service was supposed to do that, but apparently it
+    //         does not at this time.
```

Comment:
Added a bug to track whitespace contract for schemas on the service.

Reply:
OK, I'm going to remove this comment then. I don't think we should try to work around anything related to that in the meantime. There isn't much consequence for a preview. Even if we normalized here, you might have registered it elsewhere unnormalized.